packages/grafana-ui/src/components/uPlot/Plot.tsx

```diff
@@ -1,83 +1,79 @@
-    }
-
-    pluginLog('UPlotChart: init uPlot', false, 'initialized with', plotData.current, currentConfig);
-    const instance = new uPlot(currentConfig, plotData.current, canvasRef.current);
-
-    setPlotInstance(instance);
-  }, [setPlotInstance, currentConfig]);
-
+  const prevProps = usePrevious(props);
+  const prevBuilder = usePrevious(props.config);
+  const { isConfigReady, currentConfig, registerPlugin } = usePlotConfig(
+    props.width,
+    props.height,
+    props.timeZone,
+    props.config
+  );
   const getPlotInstance = useCallback(() => {
-    if (!plotInstance) {
+    if (!plotInstance.current) {
       throw new Error("Plot hasn't initialised yet");
     }
 
-    return plotInstance;
-  }, [plotInstance]);
+    return plotInstance.current;
+  }, []);
 
+  // Effect responsible for uPlot updates/initialization logic. It's performed whenever component's props have changed
   useLayoutEffect(() => {
-    plotData.current = {
-      data: props.data.frame.fields.map(f => f.values.toArray()) as AlignedData,
-      isGap: props.data.isGap,
-    };
-
-    if (plotInstance && previousConfig === props.config) {
-      updateData(props.data.frame, props.config, plotInstance, plotData.current.data);
+    // 0. Exit early if the component is not ready to initialize uPlot
+    if (!currentConfig.current || !canvasRef.current || props.width === 0 || props.height === 0) {
+      return;
     }
-  }, [props.data, props.config]);
-
-  useLayoutEffect(() => {
-    initializePlot();
-  }, [currentConfig]);
 
-  useEffect(() => {
-    const currentInstance = plotInstance;
-    return () => {
-      currentInstance?.destroy();
-    };
-  }, [plotInstance]);
+    // 1. When config is ready and there is no uPlot instance, create new uPlot and return
+    if (isConfigReady && !plotInstance.current) {
+      plotData.current = prepareData(props.data);
+      plotInstance.current = initializePlot(plotData.current, currentConfig.current, canvasRef.current);
+      return;
+    }
 
-  // When size props changed update plot size synchronously
-  useLayoutEffect(() => {
-    if (plotInstance) {
-      plotInstance.setSize({
-        width: props.width,
-        height: props.height,
+    // 2. When dimensions have changed, update uPlot size and return
+    if (currentConfig.current.width !== prevProps?.width || currentConfig.current.height !== prevProps?.height) {
+      pluginLog('uPlot core', false, 'updating size');
+      plotInstance.current!.setSize({
+        width: currentConfig.current.width,
+        height: currentConfig.current?.height,
       });
+      return;
     }
-  }, [plotInstance, props.width, props.height]);
+
+    // 3. When config or timezone has changed, re-initialize plot
+    if (isConfigReady && (props.config !== prevBuilder || prevProps.timeZone !== props.timeZone)) {
+      if (plotInstance.current) {
+        pluginLog('uPlot core', false, 'destroying instance');
+        plotInstance.current.destroy();
```

Comment:
I can see this is the only place we destroy, looks like we do not destroy on unmount

Reply:
Another great catch, thank you!